src/main.rs

```diff
@@ -155,18 +156,28 @@ fn process_file(
     let mut ser = serde_json::Serializer::with_formatter(buf, formatter);
     nb.serialize(&mut ser).map_err(|e| {
         format!(
-            "Unable to serialize notebook. Likely an intenral error: {:?}",
+            "Unable to serialize notebook. Likely an internal error: {:?}",
             e
         )
     })?;
-    let cleaned_contents = String::from_utf8(ser.into_inner()).map_err(|e| format!("{:?}", e))?;
+    let mut cleaned_contents = String::from_utf8(ser.into_inner()).map_err(|e| format!("{:?}", e))?;
 
-    if let Some(file) = output_file {
-        fs::write(&file, cleaned_contents)
-            .map_err(|e| format!("Could not write to {:?} due to {:?}", file, e))?;
+    // Check if the original content ended with a newline and the cleaned content doesn't
+    if contents.ends_with('\n') && !cleaned_contents.ends_with('\n') {
+        cleaned_contents.push('\n'); // Append a newline if necessary
+    }
+
+    if cleaned_contents != *contents {
+        if let Some(file) = output_file {
+            fs::write(&file, cleaned_contents)
+                .map_err(|e| format!("Could not write to {:?} due to {:?}", file, e))?;
+        } else {
+            println!("{}", cleaned_contents);
+        }
     } else {
-        println!("{}", cleaned_contents);
+        println!("Content unchanged. File not modified.");
```

Comment:
I'm not sure users want this written out in all cases of things not changed (especially if a file is passed in). Can we make this a `log::debug!` instead?